Let retry policy to decide about non idempotent queries

sylwiaszunejko opened the pull request:
This PR:
* Makes Conn.exec to include a flag into error that signals that query could have been actually executed, despite error been recieved.
* Picks up this error in queryExecutor.do and if query is not idempotent wrap it with user-facing error that would signal to retry policy and to user that this non idempotent query was failed but could be actually executed.

Fixes: #326, #331

dkropachev reviewed (changes requested):
On next PR let's do some heuristic from server error, take a look at [parseErrorFrame](https://github.com/scylladb/gocql/blob/fc36a673eeb41e729466b3a115cb99bba0047343/frame.go#L607-L735)

sylwiaszunejko commented on query_executor.go:
```diff
@@ -115,45 +115,61 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+					// Exit if the query was successful
+					// or no retry policy defined
+					if iter.err == nil || rt == nil {
+						return iter
```
wrap what exactly? we do not have an error here or if error is there there is no rt to consider if query should be retried or not

dkropachev replied:
My bad. Nothing to wrap up here, indeed.

sylwiaszunejko commented on query_executor.go:
```diff
@@ -115,45 +115,61 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+						// those errors represents logical errors, they should not count
+						// toward removing a node from the pool
+						selectedHost.Mark(nil)
+						return iter
```
but we don't consider rt here

dkropachev replied:
Yes, but it is not for RT only, it is only for user to know.

sylwiaszunejko replied:
but we return a iter here, should I wrap iter.err in PotentiallyExecutedNotIdempotentError?

sylwiaszunejko replied:
I pushed new version with this changed, but tbh I am not sure about that part

dkropachev commented on query_executor.go:
```diff
@@ -115,45 +115,70 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
+					iter = q.attemptQuery(ctx, qry, conn)
+					iter.host = selectedHost.Info()
+					// Update host
+					switch iter.err {
```
It won't work, we wrap it in `QueryError`, now we need to do `errors.Is` instead

sylwiaszunejko commented on conn.go:
```diff
@@ -1920,3 +1923,28 @@ type ErrSchemaMismatch struct {
+	return e.err
+}
+
+type PotentiallyExecutedNotIdempotentError struct {
```
not sure why having IsQueryLWT would be helpful?

dkropachev commented on policies.go:
```diff
@@ -250,6 +266,14 @@ func (d *DowngradingConsistencyRetryPolicy) Attempt(q RetryableQuery) bool {
+	var executedErr *QueryError
+	if errors.As(err, &executedErr) {
+		err = executedErr.err
+		if executedErr.potentiallyExecuted && !executedErr.isIdempotent {
```
Don't use them directly, have getters please.

sylwiaszunejko replied:
done

sylwiaszunejko commented:
@dkropachev ping

sylwiaszunejko commented on query_executor.go:
```diff
@@ -108,96 +108,106 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
+					potentiallyExecuted: potentiallyExecuted,
+				},
+			}, getShouldRetry(qry)
+		}
```
with this change in this case if rt = nil we would not retry, but it is not in line with current behavior, when we retry on the next host, I would change `getShouldRetry(qry)` to just `true` in first three cases

sylwiaszunejko commented:
@dkropachev I applied changes suggested by you and split everything into separate commits

dkropachev commented:
@sylwiaszunejko , I have found one more issue and fixed it in the code, pushing one more commit, please review it.

sylwiaszunejko commented:
> @sylwiaszunejko , I have found one more issue and fixed it in the code, pushing one more commit, please review it.

I incorporated this change into commits, but in this case I think simple comment to consider changing return type to RetryType instead of bool would make perfect sense and I could actually implement that myself and save you some time

dkropachev commented:
> > @sylwiaszunejko , I have found one more issue and fixed it in the code, pushing one more commit, please review it.
> 
> I incorporated this change into commits, but in this case I think simple comment to consider changing return type to RetryType instead of bool would make perfect sense and I could actually implement that myself and save you some time

Having this change in a separate commit makes it confusing because reaction on say, context.Canceled and on ErrHostDown are not preserved in between.

sylwiaszunejko commented:
@dkropachev does it look better now? Or did I miss some parts when trying to separate the changes into small separate commits?
